huaweicloud/services/live/resource_huaweicloud_live_url_validation.go

```diff
@@ -0,0 +1,248 @@
+	}
+
+	createPath := client.Endpoint + createHttpUrl
+	createPath = strings.ReplaceAll(createPath, "{project_id}", client.ProjectID)
+	createPath = fmt.Sprintf("%s?domain=%v", createPath, d.Get("domain_name"))
+
+	createOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+		JSONBody:         utils.RemoveNil(buildUrlValidationBodyParams(d)),
+	}
+
+	_, err = client.Request("PUT", createPath, &createOpt)
+	if err != nil {
+		return diag.Errorf("error creating Live URL validation: %s", err)
+	}
+
+	resourceId, err := uuid.GenerateUUID()
+	if err != nil {
+		return diag.Errorf("unable to generate ID: %s", err)
+	}
+
+	d.SetId(resourceId)
+
+	return resourceUrlValidationRead(ctx, d, meta)
+}
+
+func buildUrlValidationBodyParams(d *schema.ResourceData) map[string]interface{} {
+	params := map[string]interface{}{
+		"key":       d.Get("key"),
+		"auth_type": d.Get("auth_type"),
+		"timeout":   d.Get("timeout"),
+	}
+
+	return params
+}
+
+func resourceUrlValidationRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var (
+		cfg        = meta.(*config.Config)
+		region     = cfg.GetRegion(d)
+		domainName = d.Get("domain_name").(string)
+		getHttpUrl = "v1/{project_id}/guard/key-chain"
+	)
+
+	client, err := cfg.NewServiceClient("live", region)
+	if err != nil {
+		return diag.Errorf("error creating Live client: %s", err)
+	}
+
+	getPath := client.Endpoint + getHttpUrl
+	getPath = strings.ReplaceAll(getPath, "{project_id}", client.ProjectID)
+	getPath = fmt.Sprintf("%s?domain=%v", getPath, domainName)
+
+	getOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+	}
+
+	getResp, err := client.Request("GET", getPath, &getOpt)
+	if err != nil {
+		return common.CheckDeletedDiag(d, common.ConvertExpected400ErrInto404Err(err, "error_code", domainNameNotExistsCode),
+			"error retrieving URL validation")
+	}
+
+	getRespBody, err := utils.FlattenResponse(getResp)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	key := utils.PathSearch("key", getRespBody, "").(string)
+	if key == "" {
+		return common.CheckDeletedDiag(d, golangsdk.ErrDefault404{}, "URL validation")
+	}
+
+	mErr := multierror.Append(
+		d.Set("region", region),
+		d.Set("domain_name", domainName),
+		d.Set("key", utils.PathSearch("key", getRespBody, nil)),
+		d.Set("auth_type", utils.PathSearch("auth_type", getRespBody, nil)),
+		d.Set("timeout", utils.PathSearch("timeout", getRespBody, nil)),
+	)
+
+	if err = mErr.ErrorOrNil(); err != nil {
+		return diag.FromErr(err)
+	}
+
+	return nil
+}
+
+func resourceUrlValidationUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var (
+		cfg    = meta.(*config.Config)
+		region = cfg.GetRegion(d)
+	)
+
+	client, err := cfg.NewServiceClient("live", region)
+	if err != nil {
+		return diag.Errorf("error creating Live client: %s", err)
+	}
+
+	if d.HasChanges("key", "auth_type", "timeout") {
+		updateHttpUrl := "v1/{project_id}/guard/key-chain"
```

Comment:
Since both the create and edit operations call the same put interface, it is recommended to encapsulate this part of the logic into a separate method.